### ISSUE TITLE
Emitted chunk in plugin's transform hook lost after rebuild

### ISSUE THREAD

**Issue by edoardocavazza** [rollup/rollup#2929]
- Rollup Version: 1.15.1
- Operating System (or Browser): MacOS
- Node Version: 12.3.1

### How Do We Reproduce?

Follow the instruction in this repo: https://gitlab.com/edoardo.cavazza/rollup-watch-emit-chunk

### Expected Behavior

Emitted chunks in a plugin's `transform` hook are kept in the bundle cache.

### Actual Behavior

Emitted chunks are lost after first build. It works if the chunks are emitted in the `load` hook.

**Comment by lukastaegert** [comment]
Thanks a lot, definitely an oversight. I know we special logic to handle this case for assets, but I totally forgot about it for chunks. Hope to find some time soon to fix it.

**Comment by edoardocavazza** [comment]
Thanks @lukastaegert! In the meanwhile, I found out that using the plugin's cache to store transformations and emitted chunks, I can get it properly work re-emitting chunks:

```js
transform(code, id) {
    if (this.cache.has(id)) {
        const { chunks, result, code: previousCode } = this.cache.get(id);
        if (previousCode === code) {
            for (let file in chunks) {
                this.emitChunk(file);
            }
            return result;
        }
    }

    let result = /* transformation result... */;

    this.cache.set(id, {
        code,
        result,
        chunks,
    });

    return result;
}
```

**Comment by lukastaegert** [comment]
Fix at #2929

**Comment by edoardocavazza** [comment]
Hi @lukastaegert, I had a test with the new Rollup version, and I noticed that the incremental build works only once. At the secondo rebuild, I still have the original problem...

**Comment by lukastaegert** [comment]
Ok, so there is some improvement but not enough. I guess we need to have another go...

**Comment by lukastaegert** [comment]
Seems this is also broken for assets, so two issues that might be fixed here together.

**Comment by lukastaegert** [comment]
Proper fix that also fixes assets at #2936
